planner/core/physical_plan_test.go

```diff
@@ -253,7 +253,7 @@ func (s *testPlanSuite) TestDAGPlanBuilderJoin(c *C) {
 		},
 		{
 			sql:  "select * from t t1 join t t2 on t1.b = t2.a order by t1.a limit 1",
-			best: "IndexJoin{TableReader(Table(t))->TableReader(Table(t))}(t1.b,t2.a)->Limit",
+			best: "LeftHashJoin{TableReader(Table(t))->TableReader(Table(t))}(t1.b,t2.a)->TopN([t1.a],0,1)",
```

Comment:
is it worse than thee old result?

Reply:
Based on the current cost model, hash join is better than index join now for this case.

Reply:
Emm, based on this pr, the `expectedCnt` of hash join in this test case is not passed down. Then its cost should be no smaller than before in fact?

Reply:
Yes, cost of hash join should be larger, while cost for index join would also be larger?

Reply:
When expected count for both children of join is 10k, cost of index join is estimated to be 315k, and since index join guarantees order required by the query, cost of `Limit + IndexJoin` is 315k; while for hash join, its estimated cost is 140k, and it guarantees no order, then the final plan `TopN + HashJoin` is nearly 150k. Compared with join operator, cost of `TopN` is quite cheap in this case.

Heavy part of the index join's cost comes from line 156:
```golang
150 func (p *PhysicalIndexJoin) getCost(lCnt float64) float64 {
151     if lCnt < 1 {
152         lCnt = 1
153     }
154     cst := lCnt * netWorkFactor
155     batchSize := p.ctx.GetSessionVars().IndexJoinBatchSize
156     cst += lCnt * math.Log2(math.Min(float64(batchSize), lCnt)) * 2
157     cst += lCnt / float64(batchSize) * netWorkStartFactor
158     return cst
159 }
```
it contributes about 300k to the total cost. This line looks non-sense and is hard to understand since it is not explicitly related to any resource factor. We can change this later when refactoring these formulas.

One thing worth noting is that, for InnerJoin / SemiJoin / AntiSemiJoin, shall we multiply a factor on join's expected count and use it as the expected count of join's outer children? if we simply use MaxFloat64, the `Limit` is not reflected in the estimation of join at all. @zz-jason @winoros

Reply:
I prefer multiplying a factor, I think it will suit more cases.

Reply:
We should multiply a factor indeed, the difficult problem is how to decide that factor. Maybe use the join selectivity?

Reply:
Yes, I think it is a good choice.